File: scripts/python/lib/field_resolver.py
from .errors import AppError
from .constants import ERROR_CODES

class FieldResolver:
    def __init__(self, client, config):
        self.client = client
        self.config = config
        self.cache = {
            'fields': None
        }

    @property
    def story_points_field(self):
        return self.config.story_points_field

    def get_fields(self):
        if self.cache['fields'] is None:
            self.cache['fields'] = self.client.get_fields()
        return self.cache['fields']

    def resolve_epic_link_field(self):
        if self.config.epic_link_field:
            return self.config.epic_link_field

        fields = self.get_fields()
        match = None
        for field in fields:
            name = str(field.get('name') or '').strip().lower()
            if name in ('epic link', 'epic'):
                match = field
                break

        if not match or not match.get('id'):
            raise AppError(
                ERROR_CODES.VALIDATION,
                'Epic Link field could not be resolved. Set JIRA_EPIC_LINK_FIELD in .env.'
            )

        return match['id']

File: scripts/lib/field-resolver.js
const { AppError } = require('./errors');
const { ERROR_CODES } = require('./constants');

class FieldResolver {
  constructor(client, config) {
    this.client = client;
    this.config = config;
    this.cache = {
      fields: null
    };
  }

  get storyPointsField() {
    return this.config.storyPointsField;
  }

  async getFields() {
    if (!this.cache.fields) {
      this.cache.fields = await this.client.getFields();
    }
    return this.cache.fields;
  }

  async resolveEpicLinkField() {
    if (this.config.epicLinkField) {
      return this.config.epicLinkField;
    }

    const fields = await this.getFields();
    const match = fields.find((field) => {
      const name = String(field.name || '').trim().toLowerCase();
      return name === 'epic link' || name === 'epic';
    });

    if (!match || !match.id) {
      throw new AppError(
        ERROR_CODES.VALIDATION,
        'Epic Link field could not be resolved. Set JIRA_EPIC_LINK_FIELD in .env.'
      );
    }

    return match.id;
  }
}

module.exports = {
  FieldResolver
};
